agent: add test coverage for functions find_process and online_resources

Shensd opened the pull request:
Add test coverage for the functions `find_process` and `online_resources` in `src/sandbox.rs`.

This pull request supersedes #4084 

Fixes #4085
Fixes #4136

Signed-off-by: Jack Hance <jack.hance@ndsu.edu>

fidencio commented:
/test

liubin commented:
Hi @Shensd You need to run cargo fmt and push again.

Shensd commented:
I applied the format changes for both tests to the second commit, I figured this would be the best way to go about this as the second commit already depends on the first. If there is a better way for me to have gone about this let me know and I will make changes accordingly.

liubin commented:
> I applied the format changes for both tests to the second commit, I figured this would be the best way to go about this as the second commit already depends on the first. If there is a better way for me to have gone about this let me know and I will make changes accordingly.

Normally separated commits should be compiler-ok, I'd recommending fix the first commit and add the test case in the second commit, or squash two commits into one. Either of them, you need to do a `git rebase`.

jodh-intel commented on src/agent/src/sandbox.rs:
```diff
@@ -851,4 +852,241 @@ mod tests {
+        let (mut linux_container, _root) = create_linuxcontainer();
+
+        // add process with pid of 99
+        let test_pid = 99;
```
`libc::pid_t` is actually unsigned, so assuming black-box testing it might be good to test some boundary values too such as `std::i32::MIN`, `-1`, `0`, `1`, `std::i32::MAX` in case there are any conversions happening in `find_process()` or the functions it calls.

jodh-intel commented on src/agent/src/sandbox.rs:
```diff
@@ -851,4 +852,241 @@ mod tests {
+
+    #[tokio::test]
+    async fn test_online_resources() {
+        #[derive(Debug)]
+        struct TestFile {
+            name: String,
+            content: String,
+        }
+
+        #[derive(Debug)]
+        struct TestDirectory<'a> {
+            name: String,
+            files: &'a [TestFile],
+        }
+
+        #[derive(Debug)]
+        struct TestData<'a> {
+            directory_autogen_name: String,
+            number_autogen_directories: u32,
+
+            extra_directories: &'a [TestDirectory<'a>],
+            pattern: String,
+            to_enable: i32,
+
+            result: i32,
+        }
+
+        impl Default for TestData<'_> {
```
If you change the `struct`s to be `#[derive(Debug, Default)]`, you can delete the `impl Default` block.

jodh-intel commented on src/agent/src/sandbox.rs:
```diff
@@ -851,4 +852,241 @@ mod tests {
+            pattern: String,
+            to_enable: i32,
+
+            result: i32,
```
Could you change this to `Result<i32>` since that would:

- match exactly what `online_resources()` returns
- make it clear reading through `TestData` which tests are expected to pass/fail since you can specify a passing value as `Ok(0)`, `Ok(99)`, _etc_ and failing values as `Err(anyhow!("something bad happened"))`.

jodh-intel reviewed (commented):
Thanks @Shensd. This is quite a tricky one so great job on getting this working. Just a few comments...

fidencio reviewed (commented):
@Shensd,

Please, see the comments in the diff.

Shensd commented:
I added _most_ of the requested changes, the only one I wasn't able to fully remove was the `impl Default` for TestData in `test_online_resources`, as `#[derive(Default)]` wasn't playing nice with `Result<i32>`. 

Otherwise I have rebased and squashed the commits, I threw `cargo fmt --test` and ` cargo clippy --all-targets --all-features --release -- -D warnings` against it and it all looked good, so hopefully I covered everything.

fidencio commented:
/test

fidencio commented on src/agent/src/sandbox.rs:
```diff
@@ -847,4 +848,259 @@ mod tests {
+                1,
+            )
+            .unwrap();
+            // processes interally only have pids when manually set
```
typo: interally -> internally

fidencio commented on src/agent/src/sandbox.rs:
```diff
@@ -847,4 +848,259 @@ mod tests {
+
+        let tests = &[
+            // 4 well formed directories, request enabled 4,
+            // correct result 4 enabled, should pass
```
I really like how you describe what's expected to happen when adding the test data, kudos! :-)

fidencio reviewed (approved):
lgtm, thanks @Shensd!

fidencio commented:
/retest-devmapper-clh

fidencio commented:
/retest-arm

fidencio commented:
@jodh-intel, has @Shensd addressed all your comments?

jodh-intel reviewed (approved):
Thanks @Shensd!

lgtm
